File: backend/schemas/analysis.py
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from datetime import datetime
from uuid import UUID

class SectionScore(BaseModel):
    section_type: str
    score: int
    reasoning: Optional[str] = None

class KeywordGap(BaseModel):
    keyword: str
    frequency: Optional[int] = None

class BulletRewrite(BaseModel):
    original: Optional[str] = None
    suggested: str
    evidence_refs: Optional[List[str]] = None

class SuggestionSetResponse(BaseModel):
    section_scores: List[SectionScore]
    keyword_gaps: List[KeywordGap]
    rewrites: List[BulletRewrite]

class CareerScoreResponse(BaseModel):
    linkedin: int
    github: int
    resume: int
    overall: int
    weekly_actions: List[str]

class ResumeSuggestionResponse(BaseModel):
    match_score: int
    gap_keywords: List[KeywordGap]
    bullet_rewrites: List[BulletRewrite]
    evidence_refs: List[str]

class ScoreSnapshotItem(BaseModel):
    id: Optional[UUID] = None
    snapshotted_at: datetime
    linkedin_score: int
    github_score: int
    resume_match_score: int
    overall_score: int
    target_role: Optional[str] = None

    class Config:
        from_attributes = True

class CareerScoreHistoryResponse(BaseModel):
    target_role: Optional[str] = None
    timeframe_days: int
    total_snapshots: int
    snapshots: List[ScoreSnapshotItem]

class CareerMetricsResponse(BaseModel):
    current_overall: int
    previous_overall: int
    delta_7d: int
    current_linkedin: int
    current_github: int
    current_resume: int
    best_dimension: str
    target_role: Optional[str] = None
    market_benchmark_gap: int
    snapshotted_at: Optional[datetime] = None

class ParsedResumeResponse(BaseModel):
    name: Optional[str] = "Candidate"
    contact: Dict[str, Optional[str]] = {}
    summary: Optional[str] = ""
    experience: List[Dict[str, Any]] = []
    skills: List[str] = []
    education: List[Dict[str, Any]] = []
    certifications: List[str] = []
    raw_text: str
    word_count: int


# --- Resume Destroyer Schemas ---

class SectionBsFactor(BaseModel):
    section_name: str
    bs_factor: float
    critique: str


class RecommendedProject(BaseModel):
    repo_full_name: str
    name: str
    html_url: Optional[str] = None
    match_rationale: str
    key_technologies: List[str] = []
    suggested_bullets: List[str] = []
    stars: int = 0


class CompetitiveAnalysis(BaseModel):
    realistic_level: str
    critical_differentiators: List[str] = []
    development_priorities: List[str] = []
    market_benchmark_summary: str


class ResumeDestroyerResponse(BaseModel):
    match_score: int
    overall_bs_factor: float
    section_bs_factors: List[SectionBsFactor] = []
    critical_flaws: List[str] = []
    ats_red_flags: List[str] = []
    recommended_projects: List[RecommendedProject] = []
    bullet_rewrites: List[BulletRewrite] = []
    competitive_analysis: CompetitiveAnalysis
    gap_keywords: List[KeywordGap] = []
    evidence_refs: List[str] = []


# --- Targeted Resume CRUD Schemas ---

class TargetedResumeCreate(BaseModel):
    title: str
    target_role: str
    raw_text: str
    parsed_data: Optional[Dict[str, Any]] = None
    is_primary: Optional[bool] = False


class TargetedResumeUpdate(BaseModel):
    title: Optional[str] = None
    target_role: Optional[str] = None
    raw_text: Optional[str] = None
    parsed_data: Optional[Dict[str, Any]] = None
    is_primary: Optional[bool] = None


class TargetedResumeResponse(BaseModel):
    id: UUID
    title: str
    target_role: str
    raw_text: str
    parsed_data: Optional[Dict[str, Any]] = None
    match_score: Optional[int] = None
    bs_factor: Optional[float] = None
    last_analysis: Optional[Dict[str, Any]] = None
    is_primary: bool = False
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/analysis.ts
export interface SectionScore {
  section_type: string;
  score: number;
  reasoning?: string;
}

export interface KeywordGap {
  keyword: string;
  frequency?: number;
}

export interface BulletRewrite {
  original?: string;
  suggested: string;
  evidence_refs?: string[];
}

export interface HeadlineAlternative {
  headline: string;
  target_focus: string;
  char_count: number;
}

export interface VisualPresenceGuidance {
  photo_recommendation: string;
  banner_strategy: string;
}

export interface ExperienceRewrite {
  role_title: string;
  company?: string;
  original_snippet?: string;
  suggested_bullets: string[];
  impact_metrics: string[];
}

export interface SkillsOptimization {
  skills_to_add: string[];
  skills_to_remove: string[];
  top_pinned_skills: string[];
}

export interface LinkedInProjectRecommendation {
  repo_full_name: string;
  name: string;
  linkedin_placement: string;
  why_add: string;
  title_for_linkedin: string;
  description_snippet: string;
  skills_tags: string[];
}

export interface GrowthRoadmapPhase {
  phase: string;
  key_actions: string[];
}

export interface ContentStrategyIdea {
  topic: string;
  post_angle: string;
  target_audience: string;
  suggested_hook: string;
}

export interface SuggestionSet {
  profile_score?: number;
  headline_alternatives?: HeadlineAlternative[];
  visual_presence?: VisualPresenceGuidance;
  about_rewrite?: string;
  experience_rewrites?: ExperienceRewrite[];
  skills_optimization?: SkillsOptimization;
  recommended_projects_to_add?: LinkedInProjectRecommendation[];
  featured_section_advice?: string;
  recommendations_advice?: string;
  industry_keywords?: string[];
  competitive_analysis?: Record<string, any>;
  quick_wins?: string[];
  long_term_improvements?: string[];
  growth_roadmap?: GrowthRoadmapPhase[];
  content_ideas?: ContentStrategyIdea[];
  industry_benchmarks?: string;
  profile_completion_gaps?: string[];
  section_scores: SectionScore[];
  keyword_gaps: KeywordGap[];
  rewrites: BulletRewrite[];
}

export interface CareerScore {
  linkedin: number;
  github: number;
  resume: number;
  overall: number;
  weekly_actions: string[];
}

export interface ResumeSuggestion {
  match_score: number;
  gap_keywords: KeywordGap[];
  bullet_rewrites: BulletRewrite[];
  evidence_refs: string[];
}

export interface ScoreSnapshotItem {
  id?: string;
  snapshotted_at: string;
  linkedin_score: number;
  github_score: number;
  resume_match_score: number;
  overall_score: number;
  target_role?: string;
}

export interface CareerScoreHistory {
  target_role?: string;
  timeframe_days: number;
  total_snapshots: number;
  snapshots: ScoreSnapshotItem[];
}

export interface CareerMetrics {
  current_overall: number;
  previous_overall: number;
  delta_7d: number;
  current_linkedin: number;
  current_github: number;
  current_resume: number;
  best_dimension: string;
  target_role?: string;
  market_benchmark_gap: number;
  snapshotted_at?: string;
}
